src/3ds_ui.cpp

```diff
@@ -223,51 +253,106 @@ void CtrUi::ProcessEvents() {
 #endif
 }
 
+static __attribute__((always_inline, optimize(3))) inline u32 NDS3D_Reverse32(u32 val)
+{
+	__asm("ROR %0, %1, #24" : "=r" (val) : "r" (val));
+
+	return val;
+}
+
 void CtrUi::UpdateDisplay() {
-	main_texture->tiled = 0;
-	sf2d_fill_texture_from_RGBA8(main_texture, main_surface->pixels(),
-	                             main_surface->GetWidth(),
-	                             main_surface->GetHeight());
-	sf2d_start_frame(GFX_TOP, GFX_LEFT);
-	if (fullscreen)
-		sf2d_draw_texture_scale(main_texture, 0, 0, 1.25, 1.0);
-	else
-		sf2d_draw_texture(main_texture, 40, 0);
-	sf2d_end_frame();
-
-	sf2d_start_frame(GFX_BOTTOM, GFX_LEFT);
-	sf2d_draw_texture(keyboard_texture, 0, 0);
-	if (has_touched) {
-		// cursor
-		sf2d_draw_fill_circle(touch_x, touch_y, 5, RGBA8(0xCC, 0xCC, 0xCC, 0xAA));
-
-		u32 color = RGBA8(0xFF, 0xFF, 0xFF, 0xFF);
-		u8 width = 80;
-		u8 height = 60;
-
-		// selected button
-		u8 col = touch_x / width;
-		u8 row = touch_y / height;
-
-		// button position
-		u8 pos_x = col * width;
-		u8 pos_y = row * height;
-
-		// 0 is handled specially
+	// rotate ARGB buffer to RGBA buffer
+	// required because pixman has no fast-paths for non AXXX buffers
+	u32* line = main_buffer;
+	for (int i = 0; i <= 240; ++i) {
+		for (int j = 0; j <= 320; ++j) {
+			u32* val = line + j;
+			*val = NDS3D_Reverse32(*val);
+		}
+		line += width_pow2;
+	}
+
+	GSPGPU_FlushDataCache(main_buffer, (width_pow2*height_pow2*4));
+
+	// Using RGB8 as output format is faster and improves framerate ¯\_(ツ)_/¯
+	const u32 flags = (GX_TRANSFER_FLIP_VERT(0) |
+		GX_TRANSFER_OUT_TILED(1) |
+		GX_TRANSFER_RAW_COPY(0) |
+		GX_TRANSFER_IN_FORMAT(GX_TRANSFER_FMT_RGBA8) |
+		GX_TRANSFER_OUT_FORMAT(GX_TRANSFER_FMT_RGB8) |
+		GX_TRANSFER_SCALING(GX_TRANSFER_SCALE_NO));
+
+	// Doing this after FrameBegin corrupts the output, probably because this
+	// is asynchronous and FrameBegin will block until it finishes
```

Comment:
I have no idea what C3D_FRAME_SYNC is supposed to do, I always got corrupted output when using it. It will wait for the GPU when you pass "0", But when you pass "0" it will wait until the GPU is ready, as the C3D_FRAME_NONBLOCK doc says: "Return false instead of waiting if the GPU is busy", therefore "0" should mean: Wait.